crates/libs/windows/src/core/heap.rs

```diff
@@ -32,20 +32,42 @@ pub unsafe fn heap_free(ptr: RawPtr) {
     }
 }
 
-/// Copy a slice of `T` into a freshly allocated buffer with an additional default `T` at the end.
+/// Copy an iterator of `T` into a freshly allocated buffer with an additional default `T` at the end.
 ///
-/// Returns a pointer to the beginning of the buffer
+/// Returns a pointer to the beginning of the buffer. This pointer must be freed when done using `heap_free`.
 ///
 /// # Panics
 ///
-/// This function panics if the heap allocation fails or if the pointer returned from
-/// the heap allocation is not properly aligned to `T`.
-pub fn heap_string<T: Copy + Default + Sized>(slice: &[T]) -> *const T {
-    unsafe {
-        let buffer = heap_alloc((slice.len() + 1) * std::mem::size_of::<T>()).expect("could not allocate string") as *mut T;
-        assert!(buffer.align_offset(std::mem::align_of::<T>()) == 0, "heap allocated buffer is not properly aligned");
-        buffer.copy_from_nonoverlapping(slice.as_ptr(), slice.len());
-        buffer.add(slice.len()).write(T::default());
-        buffer
+/// This function panics if the heap allocation fails, the alignment requirements of 'T' surpass
+/// 8 (HeapAlloc's alignment) or if len is less than the number of items in the iterator.
+pub fn string_from_iter<I, T>(iter: I, len: usize) -> *const T
```

Comment:
nit: it always bothered me a bit that we're using the term `string` here when this function is more general than that. It might be nice to name it in a way that describes more closely what's actually happening. 

In fact, it might make sense for this to *only* copy an iterator and the caller is responsible for adding the trailing null byte. This function would then lose the `Default` bound and the caller would call it like so:

```rust
copy_from_iterator(self.as_bytes().iter().copied().chain(core::iter::once(0)), self.len() + 1);
```

The caller is a bit more verbose, but it's way clearer what's actually happening.